### PR TITLE
feat(plugin-meetings): separate API for enabled and muted states

### PR DESCRIPTION
**Pull request by brycetham**
# COMPLETES [SPARK-493485](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-493485)

## This pull request addresses

Updating the API to differentiate between user mute and hardware mute.

## by making the following changes

- listening to new events `UserMuteStateChange` and `SystemMuteStateChange` from webrtc-core
- updating the client local mute state to reflect both user mute and system mute
- updating the sample app to show a UI warning if the input device has been muted by the system

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

Relevant unit tests have been updated and all existing unit tests passed.

In addition, I have also tested the following cases manually using the SDK sample app:

1. SDK mutes/unmutes their own microphone/camera (user mute and system mute).
2. Host mutes SDK's microphone/camera and SDK can unmute.
3. Host mutes SDK's microphone in "moderated unmute mode", host unmutes while system is muted/unmuted.
4. SDK mutes their own microphone in "moderated unmute mode", host unmutes while system is muted/unmuted.
5. SDK changes microphone/camera devices while muted/unmuted (user mute and system mute).
6. SDK joins a meeting while muted (user mute and system mute).
7. SDK joins a meeting with "mute on entry" enabled.
8. SDK user mutes, system mutes, user unmutes, system unmutes.
9. SDK system mutes, user mutes, system unmutes, user unmutes.

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

**Comment by brycetham** [comment]
Marking this PR as ready to review, though it will need the changes from https://sqbu-github.cisco.com/WebExSquared/webrtc-media-core/pull/214 before merging.

File: packages/@webex/plugin-meetings/src/meeting/muteState.ts
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/meeting/muteState.ts:405]: would it maybe be better/safer to read `muted` property from webrtc-core instead of `systemMuted`? it would be more future proof in case anything else (not just system mute) affects the `muted` prop or in case the call to setUserMuted that's done by muteLocalStream fails for some reason. It would also be more consistent with what we do in `handleLocalStreamMuteStateChange()`

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:493]: do we have any tests that simulate system mute?

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:3044]: do we have any tests for tracks being system muted?

File: packages/@webex/plugin-meetings/src/meeting/index.ts
Comment (brycetham) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3852]: That's a good point, I'll remove the UserMuteStateChange handler for share video.

File: packages/@webex/plugin-meetings/src/meeting/muteState.ts
Comment (brycetham) [reply packages/@webex/plugin-meetings/src/meeting/muteState.ts:405]: Yeah, I'll make that change (and also update the comment).

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js
Comment (brycetham) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/muteState.js:493]: I didn't add any, but I'll take a look to see where it makes sense. Most of the tests end up checking both user mute and system mute anyways, so if anything I can add similar tests but with system mute instead of user mute.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js
Comment (brycetham) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:3044]: I'll check this too.

File: packages/@webex/media-helpers/package.json
Comment (mkesavan13) [inline packages/@webex/media-helpers/package.json:20]: We should remove this

File: packages/@webex/plugin-meetings/src/meeting/index.ts
Comment (mkesavan13) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:1299]: - From here: [feat: separate API for enabled and muted states](https://github.com/webex/webrtc-core/pull/72/files#diff-d16518f95fd5fb1272a6455c8d066afdb87254d823088f880ff255a004636467R18) I can see that the event still emits the boolean for `muted`

File: packages/@webex/plugin-meetings/src/meeting/muteState.ts
Comment (mkesavan13) [inline packages/@webex/plugin-meetings/src/meeting/muteState.ts:177]: Why wouldn't we need this anymore? Particularly when it's a publicly exposed method? We won't know which customer might still be using it

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js
Comment (mkesavan13) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:3079]: Why do these tests still say `tracks` instead of `streams`?

File: packages/@webex/media-helpers/package.json
Comment (brycetham) [reply packages/@webex/media-helpers/package.json:20]: Yep, this and the package.json file in plugin-meetings will be changed once this PR is finalized. (We didn't want to make this change until the very end to prevent it from blocking other changes in internal-media-core.)

File: packages/@webex/plugin-meetings/src/meeting/index.ts
Comment (brycetham) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:1299]: It is simply because we aren't using the parameter in this particular case. Inside `this.audio.handleLocalStreamMuteStateChange`, we manually check both `systemMuted` and `userMuted` states of the stream. The `muted` parameter here would only give us one of them... and it could be either  `systemMuted` and `userMuted`, since this handler is triggered both on `UserMuteStateChange` and `SystemMuteStateChange`.

File: packages/@webex/plugin-meetings/src/meeting/muteState.ts
Comment (brycetham) [reply packages/@webex/plugin-meetings/src/meeting/muteState.ts:177]: We pulled it out because SDK wasn't using this function anymore and the web client wasn't using it either. But if a customer might still be using it, it is an argument to put it back in.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js
Comment (brycetham) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:3079]: Probably just leftovers from when we went from track-based classes to stream-based classes and didn't notice. I'll make this change.

File: packages/@webex/plugin-meetings/src/meeting/index.ts
Comment (mkesavan13) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:1299]: Alright. Got it. Thanks @brycetham

**Review by mkesavan13 — COMMENTED** [review]
Looks good to me. Will approve after discussions with the Customers

File: docs/samples/browser-plugin-meetings/app.js
Comment (brycetham) [reply docs/samples/browser-plugin-meetings/app.js:1200]: I guess I can add it for logging purposes. The only reason why we needed it for audio is that the button label changes (we don't change the button label for video... it will always say "Toggle Video").

**Review by marcin-bazyl — APPROVED** [review]

**Review by arun3528 — APPROVED** [review]
i see @mkesavan13  review comments are address , if its urgent we can merge else wait for him

**Review by mkesavan13 — APPROVED** [review]
Looks good to me

**Pull request merged**
